miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+use itertools::chain;
+use std::collections::BTreeMap;
+
+use std::io::Result;
+use std::path::{Path, PathBuf};
+
+use duct::cmd;
+use path_macro::path;
+use walkdir::WalkDir;
+
+use crate::arg::Subcommands;
+
+const JOSH_FILTER: &str =
+    ":rev(75dd959a3a40eb5b4574f8d2e23aa6efbeb33573:prefix=src/tools/miri):/src/tools/miri";
+
+fn detect_miri_dir() -> std::io::Result<PathBuf> {
+    const MIRI_SCRIPT_ROOT_DIR: &str = env!("CARGO_MANIFEST_DIR");
+    Ok(PathBuf::from(MIRI_SCRIPT_ROOT_DIR).canonicalize()?.parent().unwrap().into())
+}
+
+fn get_current_toolchain(miri_dir: &Path) -> std::io::Result<String> {
+    let stdout = cmd!("rustup", "show", "active-toolchain").dir(miri_dir).read()?;
+    Ok(stdout.split_whitespace().next().expect("Could not obtain active Rust toolchain").into())
+}
+
+fn apply_env_variables<'a>(
+    cmd: &'a mut duct::Expression,
+    variables: &[(String, String)],
+) -> &'a mut duct::Expression {
+    for (key, value) in variables {
+        *cmd = cmd.env(key, value);
+    }
+    cmd
+}
+#[derive(Clone, Debug)]
+pub(super) struct MiriRunner<'a> {
+    /// miri_dir is based on build directory of miri-script.
+    miri_dir: PathBuf,
+    /// active_toolchain is passed as `+toolchain` argument to cargo/rustc invocations.
+    active_toolchain: String,
+    cargo_extra_flags: String,
+    command: &'a super::Subcommands,
+    /// Environment variables passed to child processes.
+    extra_flags: BTreeMap<String, String>,
+    /// Additional variables used by environment-altering commands.
+    extra_vars: Extras,
+}
+
+#[derive(Clone, Debug, Default)]
+struct Extras {
+    sysroot: PathBuf,
+    libdir: PathBuf,
+}
+
+impl Extras {
+    fn new(toolchain: &str) -> Result<Self> {
+        // Determine toolchain properties.
+        let toolchain_prop = &format!("+{toolchain}");
+        let target_output = cmd("rustc", [&toolchain_prop, "--version", "--verbose"]).read()?;
+        let host_line = target_output
+            .lines()
+            .find(|line| line.starts_with("host:"))
+            .expect("rustc output must contain host info");
+        let target = host_line.split(' ').nth(1).expect("host line does not contain host info");
+
+        let sysroot = cmd("rustc", [&toolchain_prop, "--print", "sysroot"]).read()?;
+
+        let sysroot = PathBuf::from(sysroot);
+        let libdir = path!(sysroot / "lib" / "rustlib" / target / "lib");
+        Ok(Self { libdir, sysroot })
+    }
+}
+
+impl<'a> MiriRunner<'a> {
+    pub(super) fn run(command: &'a super::Subcommands) -> Result<()> {
+        let miri_dir = detect_miri_dir()?;
+        let active_toolchain = get_current_toolchain(&miri_dir)?;
+        let config = command.get_auto_config(&miri_dir);
+        let cargo_extra_flags = std::env::var("CARGO_EXTRA_FLAGS").unwrap_or_default();
```

Comment:
I think this ends up passing an empty string to all the attempts to run `cargo`. At least it does on Linux, and so none of this actually works for me :upside_down_face: I was so excited to try this out and I definitely will when you figure out a solution to this. I thought I could hack around it quickly and I can't.

Reply:
Yeah, env variables are not propagated properly (yet). IIRC `apply_env_variables` can help with that.

Reply:
CARGO_EXTRA_FLAGS should be split into individual arguments according to shell rules, so the type here should be `Vec<String/OsString>`.